### PR TITLE
HKG: Car Port for Kia Sportage Plug-in Hybrid 2023

### PR DESCRIPTION
**Pull request by sunnyhaibin**
**Checklist**
- [x] added entry to CarInfo in selfdrive/car/*/values.py and ran `selfdrive/car/docs.py` to generate new docs
- [x] test route added to [routes.py](https://github.com/commaai/openpilot/blob/master/selfdrive/car/tests/routes.py)
- [x] route with stock system: `428412ad82894ba3|2023-10-27--15-44-21`
- [x] route with openpilot: `428412ad82894ba3|2023-12-09--17-01-47`
- [x] harness Type: [Hyundai N](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)
- [x] architecture: CAN-FD

**Link to all physical parts during installation**
  - [comma 3X](https://comma.ai/shop/comma-3x)
  - [car harness kit](https://comma.ai/shop/car-harness)
    - [Hyundai N](https://github.com/commaai/openpilot/wiki/Hyundai-Kia-Genesis#harness-guide)

Thanks to the community 2022 Kia Sportage Plug-in Hybrid owner dhimmels.

### CODE REVIEW

File: selfdrive/car/hyundai/values.py
Comment (dhimmel) [inline selfdrive/car/hyundai/values.py:2116]: Thanks @sunnyhaibin for help with the car port for my 2023 Kia Sportage PHEV X-Line Prestige.

Switching conversion here from [discord](https://discord.com/channels/469524606043160576/1167525974896803870/1180365497947148380):

> install with the following URL: `installer.comma.ai/sunnyhaibin/kia-sportage-phev-2023-port`. Once done, gather a route, and upload all rlogs and preserve the route per the instructions.

Here is a route preserved and shared on [comma connect](https://connect.comma.ai/428412ad82894ba3/1701527340790/1701527626826) with ID `428412ad82894ba3|2023-12-02--09-29-01--0`.

Also noting that I get a banner at the bottom of the screen with "CAN Error: Check Connections". I restarted the car a few times and this error message remained. Did not notice any loose or misconfigured cables.

**Comment by sshane** [comment]
@dhimmel @sunnyhaibin any reason not to combine with the current hybrid platform? We should get the HKG port into a state where this is just a fingerprint.

**Comment by sunnyhaibin** [comment]
> @dhimmel @sunnyhaibin any reason not to combine with the current hybrid platform? We should get the HKG port into a state where this is just a fingerprint.

Initially, the car seemed to do SCC from the radar, but the latest route confirmed it's doing SCC from the camera.

**Comment by sshane** [comment]
Great, then let's combine if possible

**Comment by sshane** [comment]
@dhimmel let us know when you can get a test route, just need to drive around for 10 minutes, ideally engaging with the system!

**Comment by dhimmel** [comment]
> let us know when you can get a test route, just need to drive around for 10 minutes, ideally engaging with the system!

@sshane I have [this route](https://connect.comma.ai/428412ad82894ba3/1702159307116/1702162971191) from yesterday with id `428412ad82894ba3|2023-12-09--17-01-47--32`. It might be too long (1 hour) but has several periods with OpenPilot engaged. Let me know if something shorter would be preferable for the tests?

**Comment by sshane** [comment]
I see why we have been splitting up the hybrid and plug-in hybrids because of the different masses; we should start using the info in the `hvac` FW version, which splits ICE, hybrid and plug-in hybrid soon. It seems to be available on all the variants of the Sorento too, so we can combine a couple platforms with this.

For now though, we can merge as a separate platform as that will take some large data verification.

**Comment by sshane** [comment]
@dhimmel can you put your device on WiFi and upload a short 10 minute route with some engagement? Then we can merge!

File: selfdrive/car/hyundai/interface.py
Comment (sshane) [inline selfdrive/car/hyundai/interface.py:170]:
```suggestion
    elif candidate == CAR.KIA_SPORTAGE_5TH_GEN:
```

File: selfdrive/car/hyundai/interface.py
Comment (sshane) [inline selfdrive/car/hyundai/interface.py:206]:
```suggestion
      ret.mass = {CAR.KIA_SPORTAGE_PHEV_5TH_GEN: 1923., }.get(candidate, 1767.)
```

File: selfdrive/car/hyundai/values.py
Comment (sshane) [inline selfdrive/car/hyundai/values.py:2120]: put below existing sportage hybrid

File: selfdrive/car/hyundai/values.py
Comment (sshane) [inline selfdrive/car/hyundai/values.py:2241]: same here

**Review by sshane — CHANGES_REQUESTED** [review]

File: selfdrive/car/tests/routes.py
Comment (dhimmel) [inline selfdrive/car/tests/routes.py:119]: > can you put your device on WiFi and upload a short 10 minute route with some engagement? Then we can merge!

@sshane I see you've added `428412ad82894ba3|2023-12-09--17-01-47` segment 8 for testing. Would you still like a different route? I can upload another shorter route in the next 2 days if that is helpful

**Comment by sshane** [comment]
Your car also has a very high lateral accel factor, and your car consistently overshoots the desired lateral accel, we will need to figure this out before we can merge:

image

File: selfdrive/car/tests/routes.py
Comment (sshane) [reply selfdrive/car/tests/routes.py:119]: That's fine, but can you get a route where you test the max lateral acceleration it will apply?

Comment (dhimmel) [reply selfdrive/car/tests/routes.py:119]: > can you get a route where you test the torque/max lateral acceleration it will apply

Next time I'm near highway, I will try the highway interchange or exit suggestion to get the out of torque warning. 

In the meantime, in the `428412ad82894ba3|2023-12-17--17-23-36--6` [route](https://connect.comma.ai/428412ad82894ba3/1702851815530/1702852454596) at `17:28:13 – 4`, the road took a sharp turn and the OpenPilot oversteered. Not sure if this was due to poor visibility at night with rain and oncoming headlights though.

Comment (sshane) [reply selfdrive/car/tests/routes.py:120]: Can you pull and try again? It should not be oversteering as it was.

**Comment by dhimmel** [comment]
> I have the same model and am trying to install a 3X. Can I ask where you placed the camera connector harness?

Responding to this question on discord here for easier retreivability. The following picture shows where I mounted the Comma 3x alongside the rearview mirror assembly:

image

File: selfdrive/car/hyundai/values.py
Comment (dhimmel) [inline selfdrive/car/hyundai/values.py:30]: The change in this value will end up applying to many more cars than the 2023 Sportage PHEV, right? Just curious whether the logs and your investigation uncovered a general enhancement?

File: selfdrive/car/tests/routes.py
Comment (dhimmel) [reply selfdrive/car/tests/routes.py:120]: > Can you pull and try again?

What's been holding me up is that even though the C3X is configured to use the `kia-sportage-phev-2023-port` branch, it doesn't find the latest commit even when I manually push check. It still says "up to date" but shows a stale commit 42a8a6df98c4c25c49a7ceb21676640a6ab246a6 as the current version.

Comment (dhimmel) [reply selfdrive/car/tests/routes.py:120]: > it doesn't find the latest commit even when I manually push check

Ah I think I understand the issue. I think my comma is checking the [`commaai/openpilot:kia-sportage-phev-2023-port`](https://github.com/commaai/openpilot/tree/kia-sportage-phev-2023-port) branch rather than the [`sunnyhaibin/openpilot-1:kia-sportage-phev-2023-port`](https://github.com/sunnyhaibin/openpilot-1/tree/kia-sportage-phev-2023-port) branch.

I will install from my fork that I sync and report back.

Comment (dhimmel) [reply selfdrive/car/tests/routes.py:120]: > Can you pull and try again? It should not be oversteering as it was.

Okay, here are logs based on d44e81c2727c15b2cfc43211f2b4ccf4cbe2158e: `428412ad82894ba3|2023-12-30--13-15-10--8` and [route](https://connect.comma.ai/428412ad82894ba3/1703960109006/1703960618401). I didn't reach max lateral torque on this route, so let me know if we need to capture a highway exit.

Comment (dhimmel) [reply selfdrive/car/tests/routes.py:120]: @sshane tagging for re-review with the the latest logs in this thread.

Comment (xtian) [reply selfdrive/car/tests/routes.py:120]: My experience has been that it still oversteers slightly on long curves. Installed from your fork

File: selfdrive/car/hyundai/values.py
Comment (sshane) [reply selfdrive/car/hyundai/values.py:30]: Yes, I only did that to test if it fixed the problems

**Comment by Wafflezzbutt** [comment]
I have a 2024 Sportage PHEV that has very few changes from the 2023. If there is anything I can contribute to add it to the compatibility list

`'carFw': [{'address': 2000,
                          'brand': 'hyundai',
                          'bus': 0,
                          'ecu': 'fwdRadar',
                          'fwVersion': b'\xf1\x00NQ5__               1.01 1.03 99110-CH000         ',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 2008,
                          'subAddress': 0},
                         {'address': 1988,
                          'brand': 'hyundai',
                          'bus': 0,
                          'ecu': 'fwdCamera',
                          'fwVersion': b'\xf1\x00NQ5 FR_CMR AT USA LHD 1.00 1.00 99211-P1070 690',
                          'logging': False,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 1996,
                          'subAddress': 0},
                         {'address': 1975,
                          'brand': 'hyundai',
                          'bus': 0,
                          'ecu': 'cornerRadar',
                          'fwVersion': b'\xf1\x00NQ5 BCW RR 1.00 , 1.00 (\x81\x98\x83#\x02u\x06Y',
                          'logging': True,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 1983,
                          'subAddress': 0},
                         {'address': 1971,
                          'brand': 'hyundai',
                          'bus': 0,
                          'ecu': 'hvac',
                          'fwVersion': b'\xf1\x00NQ5PH 97255-CH100UNIT-HEATER CONTROL 1.03NQ5PHEV DATC 1.2 1.01',
                          'logging': True,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x00'],
                          'responseAddress': 1979,
                          'subAddress': 0},
                         {'address': 1969,
                          'brand': 'hyundai',
                          'bus': 0,
                          'ecu': 'parkingAdas',
                          'fwVersion': b'\xf1\x10NQ5  ADAS_PRK ANL 1.01 1.06 99910-CH330',
                          'logging': True,
                          'obdMultiplexing': True,
                          'request': [b'"\xf1\x10'],
                          'responseAddress': 1977,
                          'subAddress': 0}],`

**Comment by Wafflezzbutt** [comment]
https://useradmin.comma.ai/?onebox=ad5d1c08bcc42d1a%7C2024-02-15--17-19-23

I do feel like it overcorrects sometimes. It seems to "lean left" especially around turns, and then over correct back

File: selfdrive/car/tests/routes.py
Comment (dhimmel) [reply selfdrive/car/tests/routes.py:120]: Okay, so @xtian and I find that there is a bit of oversteering on sharp turns with the 2023 Kia Sportage PHEV. 

@Wafflezzbutt [reports](https://github.com/commaai/openpilot/pull/30575#issuecomment-1949546120) the same on the 2024 Kia Sportage PHEV (with additional fingerprints added).

> I do feel like it overcorrects sometimes. It seems to "lean left" especially around turns, and then over correct back

So @sshane / @sunnyhaibin my questions are:

1. is the oversteering issue also present with Hybrid and Gas models of the Kia Sportage 2023+? If so I wonder if the best approach is to merge and then fix all models.
2. Independent of 1, how do we go about fixing the oversteeting?

**Comment by xtian** [comment]
Is this PR blocked on needing additional route data?

**Comment by Wafflezzbutt** [comment]
While The turning radius issue seems to be less of a problem since the 0.9.6 release, I think there may be an issue with how heavy the car is. The PHEV model is 5300 lbs compared to the 3500 on the regular sportage. In experimental mode it seems to have trouble braking if there is a downward slope and it occasionally gets too close to the car in front of it when coming to a complete stop

Here is a drive today where this should be obvious:
https://useradmin.comma.ai/?onebox=ad5d1c08bcc42d1a%7C2024-03-12--17-40-15

Especially:
Segment 3 
Segment 6
Segment 10

**Comment by xtian** [comment]
I get a “Driver’s grasp not detected” alert from the stock ACC relatively often while using @dhimmel’s fork. Are others experiencing this?

**Comment by xtian** [comment]
Is there anything I can do to help this get merged?

**Comment by sshane** [comment]
We've moved the car interfacing code to our [`opendbc`](https://github.com/commaai/opendbc) repository, which is now the new home for car ports and fingerprints. Please re-open your pull request against [`opendbc`](https://github.com/commaai/opendbc) at your convenience by using this command below. This will transform all changes under `selfdrive/car/` to `opendbc_repo/opendbc/car/`. Make sure you have initialized submodules beforehand and have checked out the latest `opendbc` commit.

```bash
PR_NUMBER=33045
curl -L https://github.com/commaai/openpilot/pull/$PR_NUMBER.patch | sed -e 's/selfdrive\/car/opendbc_repo\/opendbc\/car/g' | git apply -v --reject
```

Simply replace the PR number with your own. Once done, add the files, fix any conflicts, and open a new PR. Alternatively, you may start a new PR from scratch if that is easier for you.

**Comment by dhimmel** [comment]
@xtian or @Wafflezzbutt would either of you like to pick up the mantle and migrate the PR to `opendbc`? I'd usually jump at the opportunity to contribute to such an impactfull open source library, but am a bit pressed for time these days. Happy to support with any evaluation and test drive work.

**Comment by xtian** [comment]
I might be able to, but first I'd need to know why this PR couldn't be merged after almost nine months.

**Comment by dhimmel** [comment]
@xtian my understanding as an onlooker and not a maintainer is that this PR languished because of the oversteering issue and that the original author @sunnyhaibin was occupied with other work. But it looks like @sunnyhaibin has opened https://github.com/commaai/opendbc/pull/1138 and https://github.com/commaai/openpilot/pull/33341, so we should direct any assistance efforts to those PRs.

**Pull request closed without merging**
